runtime/src/bank.rs

```diff
@@ -1907,85 +1958,114 @@ impl Bank {
+                        vote_state,
+                        (stake_pubkey, (stake_state, mut stake_account)),
+                    )| {
+                        let reward_calc_tracer = reward_calc_tracer.as_ref().map(|outer| {
+                            // inner
+                            move |inner_event: &_| {
+                                outer(&RewardCalculationEvent::Staking(&stake_pubkey, inner_event))
+                            }
+                        });
+                        let redeemed = stake_state::redeem_rewards(
+                            rewarded_epoch,
+                            stake_state,
+                            &mut stake_account,
+                            &vote_state,
+                            &point_value,
+                            Some(&stake_history),
+                            &reward_calc_tracer,
+                            fix_stake_deactivate,
+                        );
+                        if let Ok((stakers_reward, voters_reward)) = redeemed {
+                            // track voter rewards
+                            if voters_reward > 0 {
+                                if let Some((_vote_account, _commission, vote_rewards_sum)) =
+                                    vote_account_rewards.get_mut(&vote_pubkey).as_deref_mut()
+                                {
+                                    *vote_rewards_sum += voters_reward
+                                }
+                            }
+
+                            // store stake account even if stakers_reward is 0
+                            // because credits observed has changed
+                            self.store_account(&stake_pubkey, &stake_account);
+
+                            if stakers_reward > 0 {
+                                return Some((
+                                    stake_pubkey,
+                                    RewardInfo {
+                                        reward_type: RewardType::Staking,
+                                        lamports: stakers_reward as i64,
+                                        post_balance: stake_account.lamports,
+                                        commission: Some(vote_state.commission),
+                                    },
+                                ));
+                            }
+                        } else {
+                            debug!(
+                                "stake_state::redeem_rewards() failed for {}: {:?}",
+                                stake_pubkey, redeemed
+                            );
+                        }
+                        None
+                    },
+                )
+                .collect()
+        });
+
+        let mut vote_rewards = vote_account_rewards
+            .into_iter()
+            .filter_map(
+                |(vote_pubkey, (mut vote_account, commission, vote_rewards))| {
+                    if vote_rewards == 0 {
```

Comment:
hmm, sadly, `voters_reward > 0` can't be true substitute for the old `vote_account_changed` flag.

this new condition would create differing bank hash (due to the fact certain kind of vote accounts are stored or no... so, this is not inflation rewards discrepancy per se), if i read the code correctly..

this is because previous code always `store_account()`s vote accounts even if (total) `vote_rewards` is 0: https://github.com/solana-labs/solana/pull/20339/files#r719652698

the new code doesn't `store_account()` vote accounts if (total) `vote_rewards` is 0.

Reply:
I checked that vote_account isn't modified besides lamport changes so I don't think we need the store account here. But it doesn't seem like it would hurt, I'll add it back

Reply:
The hashing will look at all accounts, so even if it's the same value, the hash will be different if the account update is present in the new slot or not.